File: 01-GeekBand算法原理与实践/python/coinChange.py
"""
 * @file     coinChange.py
 * @brief    
 * @author   YongDu
 * @date     2021-07-05
"""
import math
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        result = [math.inf for _ in range(amount + 1)]
        result[0] = 0

        for i in range(1, amount + 1):
            for coin in coins:
                if i >= coin:
                    result[i] = min(result[i-coin] + 1, result[i])

        return result[-1] if result[-1] != math.inf else -1

File: 01-GeekBand算法原理与实践/cpp/coinChange.cpp
/**
 * @file     coinChange.cpp
 * @brief
 * @author   YongDu
 * @date     2021-07-05
 */
class Solution {
  public:
    int coinChange(vector<int> &coins, int amount) {
        vector<int> result(amount + 1, INT_MAX);
        result[0] = 0;
        int coinNums = coins.size();

        for (int i = 1; i <= amount; ++i) {
            for (int j = 0; j < coinNums; ++j) {
                if (i - coins[j] >= 0 && result[i - coins[j]] != INT_MAX) {
                    result[i] = std::min(result[i], result[i - coins[j]] + 1);
                }
            }
        }
        if (INT_MAX == result[amount]) {
            return -1;
        }
        return result[amount];
    }
};
